components/merkledb/src/views/owning_iter.rs

```diff
@@ -0,0 +1,337 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{marker::PhantomPinned, mem, pin::Pin};
+
+use crate::{
+    db::Iterator,
+    views::{RawAccess, View},
+};
+
+#[derive(Debug)]
+struct Inner<T: RawAccess> {
+    // Safety: Not accessed during iteration.
+    view: View<T>,
+    iter: Option<Box<dyn Iterator>>,
+    _pin: PhantomPinned,
+}
+
+impl<T: RawAccess> Inner<T> {
+    #[allow(unsafe_code)]
+    fn new(view: View<T>, from: &[u8]) -> Pin<Box<Self>> {
+        let mut boxed = Box::pin(Self {
+            view,
+            iter: None,
+            _pin: PhantomPinned,
+        });
+
+        unsafe {
+            // Elongate the iterator lifetime to `'static`.
+            //
+            // SAFETY:
+            // `T` may not have static lifetime (indeed, `T == &Fork` or `T == &dyn Snapshot`
+            // are two frequent use cases), but since `Inner` contains `T` as the type param,
+            // `iter` is always valid during the lifetime of the `Inner` instance.
+            let iter: Box<dyn Iterator> = mem::transmute(boxed.view.iter_bytes(from));
```

Comment:
It's possible to get rid of `transmute` here in favor of reference lifetime elongation:

```rust
struct Inner<'a, T: RawAccess + 'a> {
    view: View<T>,
    iter: Option<Box<dyn Iterator + 'a>>,
    _pin: PhantomPinned,
}

let view_ptr = &boxed.view as *const View<T>;
let view_ref: &'a View<T> = unsafe { &*view_ptr };
let iter: Box<dyn Iterator + 'a> = View::iter_bytes(view_ref, from);
```

In code using the iterator, the lifetime `'a` would be automatically set by the compiler. For example, the iterator in `Group` would be redefined as

```rust
pub fn keys<'a>(&self) -> Keys<'a, T::Base, K>
``` 

Not sure whether this complication is warranted though.